src/memoryprofiler.js

```diff
@@ -0,0 +1,379 @@
+    // Also track if this was a _malloc performed at preRun time.
+    if (!this.pagePreRunIsFinished) this.preRunMallocs[ptr] = size;
+
+    if (this.allocateStatistics) {
+      var loc = new Error().stack.toString();
+      var str = loc;
+      if (!this.allocationSiteStatistics[str]) this.allocationSiteStatistics[str] = 1;
+      else ++this.allocationSiteStatistics[str];
+      this.allocationSitePtrs[ptr] = loc;
+    }
+
+    // If this is a large enough allocation, track its detailed callstack info.
+    if (size > this.trackedCallstackMinSizeBytes) {
+      // Get the caller function as string.
+      var loc = new Error().stack.toString();
+      var nl = loc.indexOf('\n')+1;
+      loc = loc.substr(nl);
+      loc = loc.replace(/\n/g, '<br />');
+      var caller = loc;
+      this.callstackOfAllocatedPtr[ptr] = caller;
+      if (this.allocatedBytesPerCallstack[caller] > 0) this.allocatedBytesPerCallstack[caller] += size;
+      else this.allocatedBytesPerCallstack[caller] = size;
+    }
+  },
+
+  onFree: function onFree(ptr) {
+    if (!ptr) return;
+
+    // Decrement global stats.
+    var sz = this.allocatedPtrSizes[ptr];
+    this.totalMemoryAllocated -= sz;
+    delete this.allocatedPtrSizes[ptr];
+    delete this.preRunMallocs[ptr]; // Also free if this happened to be a _malloc performed at preRun time.
+    this.stackTopWatermark = Math.max(this.stackTopWatermark, STACKTOP);
+
+    if (this.allocateStatistics) {
+      var loc = this.allocationSitePtrs[ptr];
+      if (loc) {
+        var str = loc;
+        --this.allocationSiteStatistics[str];
+      }
+      this.allocationSitePtrs[ptr] = null;
+    }
+
+    // Decrement per-alloc stats if this was a large allocation.
+    if (sz > this.trackedCallstackMinSizeBytes) {
+      var caller = this.callstackOfAllocatedPtr[ptr];
+      delete this.callstackOfAllocatedPtr[ptr];
+      this.allocatedBytesPerCallstack[caller] -= sz;
+      if (this.allocatedBytesPerCallstack[caller] <= 0) delete this.allocatedBytesPerCallstack[caller];
+    }
+    ++this.totalTimesFreeCalled;
+  },
+
+  onRealloc: function onRealloc(oldAddress, newAddress, size) {
+    onFree(oldAddress);
+    onMalloc(newAddress, size);
+  },
+
+  onPreloadComplete: function onPreloadComplete() {
+    this.pagePreRunIsFinished = true;
+  },
+
+  // Installs startup hook and periodic UI update timer.
+  initialize: function initialize() {
+    // Inject the memoryprofiler hooks.
+    Module['onMalloc'] = function onMalloc(ptr, size) { emscripten_memoryprofiler.onMalloc(ptr, size); };
+    Module['onRealloc'] = function onRealloc(oldAddress, newAddress, size) { emscripten_memoryprofiler.onRealloc(oldAddress, newAddress, size); };
+    Module['onFree'] = function onFree(ptr) { emscripten_memoryprofiler.onFree(ptr); };
+
+    // Add a tracking mechanism to detect when VFS loading is complete.
+    Module['preRun'].push(function() { emscripten_memoryprofiler.onPreloadComplete(); });
```

Comment:
That is most likely redundant now. I don't have a test case to verify, and it is a long time since I added that line. Removed it now and updated the PR.